clients/src/main/java/org/apache/kafka/common/security/JaasConfig.java

```diff
@@ -50,12 +50,24 @@
     private final List<AppConfigurationEntry> configEntries;
 
     public JaasConfig(String loginContextName, String jaasConfigParams) {
+        // All characters except space, comment, quote, equal and semicolon are considered to be alphabetic.
+        // Tokenizer rules:
+        // 1. All bytes from 0 to 32 ({@code ' '}) are considered to be whitespace.
+        // 2. {@code '/'} (47) is a comment character. '//', '/*', '*/' are also allowed.
+        // 3. Single quote ({@code '\u005C''}, 39) and double quote ({@code '"'}, 34) are considered to be quote.
+        // 4. Ends of lines are treated as white space, not as separate tokens.
         StreamTokenizer tokenizer = new StreamTokenizer(new StringReader(jaasConfigParams));
-        tokenizer.slashSlashComments(true);
-        tokenizer.slashStarComments(true);
-        tokenizer.wordChars('-', '-');
-        tokenizer.wordChars('_', '_');
-        tokenizer.wordChars('$', '$');
+        tokenizer.resetSyntax();            // Reset the default configuration.
+        tokenizer.wordChars(32, 128);       // All characters in [32, 128] are allowed.
+        tokenizer.wordChars(128 + 32, 255); // All characters in [160, 255] are allowed.
+        tokenizer.ordinaryChar(';');        // ';' is treated as a reserved word.
+        tokenizer.ordinaryChar('=');        // '=' is treated as a reserved word.
+        tokenizer.whitespaceChars(0, ' ');  // All characters in [0, 32] (including ' ') are treated as space character.
+        tokenizer.commentChar('/');         // '/' is treated as a comment character.
+        tokenizer.quoteChar('"');           // '"' is treated as a quote.
+        tokenizer.quoteChar('\'');          // ''' is treated as a quote.
+        tokenizer.slashSlashComments(true); // Allow '//' comments.
+        tokenizer.slashStarComments(true);  // Allow '/*', '*/' comments.
```

Comment:
@dongjinleekr Thanks for the PR, can we add some tests to ensure that the syntax is identical to that of Java's standard file-based JAAS `Configuration` parsing?

Reply:
Hi @rajinisivaram,

I searched the definition of Java's standard file-based JAAS `Configuration` but could not find one. But, after reviewing [the related `LoginModule` implementations](https://docs.oracle.com/javase/10/security/appendix-b-jaas-login-configuration-file.htm) ([#1](https://github.com/openjdk/jdk/blob/jdk-9%2B181/jdk/src/jdk.security.auth/share/classes/com/sun/security/auth/module/KeyStoreLoginModule.java), [#2](https://github.com/openjdk/jdk/blob/jdk-9%2B181/jdk/src/jdk.security.auth/share/classes/com/sun/security/auth/module/Krb5LoginModule.java)) and [example](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/security-reference/content/kerberos_nonambari_create_jaas_configuration_files.html), what we need to test seems almost clear:

- It should be able to parse a string (including non-reserved symbols), number, boolean, asterisk, canonical java class name, and URL; regardless it is quoted or not.
- **We should remove the extended ASCII codes (i.e., [128, 256)) from the PR.** That is, we should allow 7-bit alphabets and symbols only.

Is this okay? Then, I will proceed to update the tests accordingly.

Reply:
@dongjinleekr I think the base implementation in Java comes from `sun.security.provider.ConfigFile`, so maybe worth checking that. For both the positive and negative cases, we can create an example for test, parse it using Kafka's `JaasConfig` and also write to a temp file and parse by loading standard `Configuration`, verify that both have the same behaviour. Obviously we can't add exhaustive tests, the various cases discussed in this PR would be good enough.

Reply:
Hi @rajinisivaram,

I reviewed the implementation of `sun.security.provider.ConfigFile` [here](https://github.com/openjdk/jdk/blob/jdk-9%2B181/jdk/src/java.base/share/classes/sun/security/provider/ConfigFile.java#L413) and found the following:

- `ConfigFile` treats only '\*', '\_', '-', '$' as word characters; that is, the numbers or other symbols like '%', '^' are not allowed by default - if the user needs to use these symbols, they should use quotes. So, **the problem described in the Jira issue is actually not a problem, but rather a documentation issue.** Instead, we need to add '*' as a word character. (Currently, it is not.) Obviously, it is a bug.
- It seems like we don't need to define the ordinary characters (like ';', '='), spaces, comment characters, and quote characters explicitly, like `ConfigFile` does.
- It would be better to improve `JaasContextTest` by not only checking it works like the documentation, but also checking it works like the original, `ConfigFile` implementation.

I greatly appreciate you for guiding the right direction. I am now updating the PR now. Thanks again and stay tuned! :smiley:

Reply:
Hi @rajinisivaram,

I reviewed the implementation of `sun.security.provider.ConfigFile` [here](https://github.com/openjdk/jdk/blob/jdk-9%2B181/jdk/src/java.base/share/classes/sun/security/provider/ConfigFile.java#L413) and found the following:

- `ConfigFile` treats only `*`, `_`, `-`, `$` as word characters; that is, the numbers or other symbols like `%`, `^` are not allowed by default - if the user needs to use these symbols, they should use quotes. So, the problem described in the Jira issue is actually not a problem, but rather a documentation issue. Instead, we need to add `*` as a word character. Currently, it is not and obviously, it is a bug.
- It seems like we don't need to define the ordinary characters (like `;`, `=`), spaces, comment characters, and quote characters explicitly, like `ConfigFile` does.
- It would be better to improve `JaasContextTest` by not only checking it works like the documentation, but also checking it works like the original, `ConfigFile` implementation. (I am trying to, but can't certain yet.)

I greatly appreciate you for guiding the right direction. I am now updating the PR now. Thanks again and stay tuned! :smiley: